File: algo/lc.48.cc
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int s = matrix.size();
        
        // frist pass: transpose
        for (int i = 0; i < s; i++) {
            for (int j = 0; j < i; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }
        
        // second pass: flipping left-right
        for (int i = 0; i < s; i++) {
            for (int j = 0; j < s/2; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[i][s-1-j];
                matrix[i][s-1-j] = tmp;
            }
        }
    }
};

File: algo/lc.48.py
class Solution:
    def transpose(self, matrix: List[List[int]]) -> None:
        '''
        in-place transpose
        '''
        N = len(matrix)
        for i in range(N):
            for j in range(i):
                tmp = matrix[i][j]
                matrix[i][j] = matrix[j][i]
                matrix[j][i] = tmp
    def lrflip(self, matrix: List[List[int]]) -> None:
        N = len(matrix)
        for i in range(N):
            for j in range(N // 2):
                tmp = matrix[i][j]
                matrix[i][j] = matrix[i][N-j-1]
                matrix[i][N-j-1] = tmp
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.transpose(matrix)
        self.lrflip(matrix)
